sql/catalog: unify the descs.Collection resolution-by-ID API

thoszhang opened the pull request:
See commits for details. This PR mostly standardizes the `ByID` methods on `descs.Collection` for the specific descriptor types.

Closes #57539.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57343)
<!-- Reviewable:end -->

Commit:
sql/catalog: unify descs.Collection ID resolution, standardize methods

This commit unifies the implementations of all the `descs.Collection`
resolution-by-ID methods, and increases their consistency in respecting
lookup flags for non-public descriptors. It also adds methods with
standardized names and signatures: each descriptor type now has mutable
and immutable variants for resolution by ID. All the pre-existing
methods still exist, but are marked for deprecation.

Release note: None

Commit:
sql/catalog: rename/replace descs.Collection methods for consistency

This commit performs a few trivial replacements of `descs.Collection`
methods with their new names:

- `GetTableVersionByID` -> `GetImmutableTableByID`
- `GetTypeVersionByID` -> `GetImmutableTypeByID`
- `ResolveSchemaByID` -> `GetImmutableSchemaByID` (which also takes an
  explicit flag)
- `GetDatabaseVersionByID` -> `GetImmutableDatabaseByID`

All the `Required` values have also been removed from the flags passed
into these methods, since they unconditionally return an error when the
descriptor does not exist.

Release note: None

lucy-zhang commented:
I think as a first pass this PR does most of what we want, API-wise. There are still some loose ends: There are a few deprecated methods left on the `Collection` (`GetMutableTableVersionByID`) whose existing usages need to be replaced, and we still don't have `GetImmutableObjectByName`, etc., but I think it's a reasonable stopping point.

I'm open to renaming the old `GetMutableTableVersionByID` and `GetMutableTypeVersionByID` methods to names that make more sense before we mostly set this part of the interface overhaul aside.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Sorry this took so long. Once things start getting hidden in the reviewable UI I lose them. 

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57343#-:-MR1-N0p61E5CdJMYikV:b-dcep4t)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/catalog/descs/collection.go, line 899 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57343#-MR0zyMs1XOLqBjktUKI:-MR0zyMs1XOLqBjktUKJ:bbtffyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0284ce894ea079da80a2d1175239a454bd2078e7/pkg/sql/catalog/descs/collection.go#L899)):*
> ```
> 		// TODO (lucy): Consider how to avoid having to generate user-facing errors
> 		// here. At the very least, this is inconsistent with the name resolution
> 		// methods, which just return internal errors from filterDescriptorState.
> ```
Can you help me better understand what this would take?


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
Also removed the explicit `Required` flags from the `ByID` method calls, since we ignore that flag.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57343#-:-MR2sS4DDKsO7oq7cIvr:b9sczch)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)

___
*[pkg/sql/catalog/descs/collection.go, line 899 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57343#-MR0zyMs1XOLqBjktUKI:-MR2lEwD6AxL-08mWtdz:b-3o2wm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0284ce894ea079da80a2d1175239a454bd2078e7/pkg/sql/catalog/descs/collection.go#L899)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 		// TODO (lucy): Consider how to avoid having to generate user-facing errors
> 		// here. At the very least, this is inconsistent with the name resolution
> 		// methods, which just return internal errors from filterDescriptorState.
> ```
Can you help me better understand what this would take?
</blockquote></details>

I deleted this comment, since I'm not convinced that this is a problem, or at least not the most important problem.

I just fixed a bug in this PR that caused `TestInvalidOwnedDescriptorsAreDroppable` to fail, due to us checking specifically for `catalog.ErrDescriptorNotFound` (which we used to directly return out of the physical accessor, or something, but now gets discarded in favor of the various user-facing errors in `sqlerrors`). I fixed it by having us check for `pgcode.UndefinedTable` instead, but I don't know that this is a good idea.

The point is, it seems potentially useful to have a reliable sentinel error indicating that no such descriptor exists with the given name/ID. We didn't really do this starting with the name resolution APIs, so I don't want to do it in this PR, but I opened #59027 with some thoughts.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
cc @postamar - let's get you tagged on reviews like this. 

> Reviewed 1 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57343#-:-MR35w-n62bAu_i9yJ_e:bksmuia)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 3 files at r2, 14 of 14 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57343#-:-MR36q67CgkBbQsCEJXr:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/catalog/descs/collection.go, line 899 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57343#-MR0zyMs1XOLqBjktUKI:-MR36LfBCUTzmVGrYODT:bryibot) ([raw file](https://github.com/cockroachdb/cockroach/blob/0284ce894ea079da80a2d1175239a454bd2078e7/pkg/sql/catalog/descs/collection.go#L899)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I deleted this comment, since I'm not convinced that this is a problem, or at least not the most important problem.

I just fixed a bug in this PR that caused `TestInvalidOwnedDescriptorsAreDroppable` to fail, due to us checking specifically for `catalog.ErrDescriptorNotFound` (which we used to directly return out of the physical accessor, or something, but now gets discarded in favor of the various user-facing errors in `sqlerrors`). I fixed it by having us check for `pgcode.UndefinedTable` instead, but I don't know that this is a good idea.

The point is, it seems potentially useful to have a reliable sentinel error indicating that no such descriptor exists with the given name/ID. We didn't really do this starting with the name resolution APIs, so I don't want to do it in this PR, but I opened #59027 with some thoughts.
</blockquote></details>

I think the answer here should be to leverage [`errors.Mark`](https://godoc.org/github.com/cockroachdb/errors#Mark) to make sure that the meaningful, identifiable error object propagates. Using the pgcodes doesn't make me happy. 


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57343#-:-MR3940QE2jsBVdM5N2y:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/catalog/descs/collection.go, line 899 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57343#-MR0zyMs1XOLqBjktUKI:-MR36z6qEZsTobZng4q9:b-6reaj9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0284ce894ea079da80a2d1175239a454bd2078e7/pkg/sql/catalog/descs/collection.go#L899)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think the answer here should be to leverage [`errors.Mark`](https://godoc.org/github.com/cockroachdb/errors#Mark) to make sure that the meaningful, identifiable error object propagates. Using the pgcodes doesn't make me happy. 
</blockquote></details>

FWIW, we already check for that pgcode here, where the logic is basically the same:

https://github.com/cockroachdb/cockroach/blob/4fcc73a79c1cd23d720e61fcd727d39a660982e5/pkg/sql/drop_sequence.go#L170-L174

So I just copied it. But I don't like it either.

We should try to fix the errors soon, though I fear that it won't be that easy since the callers probably don't actually handle errors (or what they indicate) in a very principled way and we just rely on a lot of accidental behavior. Maybe we can do that after finally mocking out the physical access dependencies like we've been meaning to, though that in itself is not necessarily going to help us do the right things.


<!-- Sent from Reviewable.io -->
